Allow random-1.3

bgamari opened the pull request:
Implement `uniformByteArrayM` in terms of `uniformByteArray`. I'll admit, this seems almost too easy to be correct so do let me know if there is something I have missed.

Fixes #54.

bgamari committed:
Allow random-1.3

Implement `uniformByteArrayM` in terms of `uniformByteArray`.

Fixes #54.

bgamari commented:
Pinging @lehins.

byorgey commented:
Wow, that does seem almost too easy.  I had been working on upgrading but my code was way more complicated (see https://github.com/haskell/random/issues/179 ).  I will take a look, but would also love any insights from @lehins .

byorgey commented:
Hmm, I looked into this a bit more, and it seems very plausible!  I'm glad I didn't work on my overly complex version any more.

I don't know why the CI keeps being cancelled.  I try re-running it and after a few seconds it says all the checks were cancelled by me.

RyanGlScott commented:
Your CI is using `ubuntu-20.04` runners, which is scheduled for removal (and currently not available due to a brownout—see https://github.com/actions/runner-images/issues/11101). Regenerating the `.github/workflows/haskell-ci.yml` file with a newer version of `haskell-ci` would likely fix this.

lehins commented on Control/Monad/Trans/Random/Lazy.hs:
```diff
@@ -283,6 +283,9 @@ instance (Monad m, RandomGen g) => StatefulGen (RandGen g) (RandT g m) where
   uniformWord32 = applyRandT genWord32
   uniformWord64 = applyRandT genWord64
   uniformShortByteString n = applyRandT (genShortByteString n)
+#if MIN_VERSION_random(1,3,0)
+  uniformByteArrayM pinned sz = applyRandT $ uniformByteArray pinned sz
+#endif
```
I'd actually recommend switching to default implementation for `uniformShortByteString`, since it is now deprecated (it uses the underneath `uniformByteArrayM` for default implementation from 1.3.0 onwards) and will be removed in some future . This will improve future compatibility.
```suggestion
#if MIN_VERSION_random(1,3,0)
  uniformByteArrayM pinned sz = applyRandT $ uniformByteArray pinned sz
#else
  uniformShortByteString n = applyRandT (genShortByteString n)
#endif
```

lehins commented on Control/Monad/Trans/Random/Strict.hs:
```diff
@@ -290,6 +290,9 @@ instance (Monad m, RandomGen g) => StatefulGen (RandGen g) (RandT g m) where
   uniformWord32 = applyRandT genWord32
   uniformWord64 = applyRandT genWord64
   uniformShortByteString n = applyRandT (genShortByteString n)
+#if MIN_VERSION_random(1,3,0)
+  uniformByteArrayM pinned sz = applyRandT $ uniformByteArray pinned sz
```
```suggestion
#if MIN_VERSION_random(1,3,0)
  uniformByteArrayM pinned sz = applyRandT $ uniformByteArray pinned sz
#else
  uniformShortByteString n = applyRandT (genShortByteString n)
```

lehins reviewed (approved):
That is exactly it. I intended for it to have a simple implementation. Thank you @bgamari for submitting this PR. Sorry @byorgey I totally forgot that you were asking about some help defining implementation for `uniformByteArrayM` in `MonadRandom`

byorgey committed:
Apply suggestions from code review

Implement either `uniformByteArrayM` or `uniformShortByteString` depending on the version of `random`, but not both

Co-authored-by: Alexey Kuleshevich <lehins@yandex.ru>
